src/compiler/checker.ts

```diff
@@ -7646,11 +7646,16 @@ namespace ts {
                     resolvedSymbol.exports = createSymbolTable();
                 }
                 (resolvedSymbol || symbol).exports!.forEach((s, name) => {
-                    if (members.has(name)) {
-                        const exportedMember = exportedType.members.get(name)!;
-                        const union = createSymbol(s.flags | exportedMember.flags, name);
-                        union.type = getUnionType([getTypeOfSymbol(s), getTypeOfSymbol(exportedMember)]);
-                        members.set(name, union);
+                    const exportedMember = members.get(name)!;
+                    if (exportedMember && exportedMember !== s) {
+                        if (s.flags & SymbolFlags.Value) {
+                            const union = createSymbol(s.flags | exportedMember.flags, name);
+                            union.type = getUnionType([getTypeOfSymbol(s), getTypeOfSymbol(exportedMember)]);

```

Comment:
The problem is that `s` can itself be a type symbol, whether by module augmentation or JS `@typedef`, so `getTypeOfSymbol(s)` would return `errorType` much of the time. In other cases, `exportedMember === s` and so this only served to needlessly create a transient symbol missing `declarations`. (That’s why you see a bunch of symbol baselines gain a value declaration.)

But, this is probably not quite right for value symbols that should be allowed to merge, like a function from JS merged with a namespace from a module augmentation... The only test that demonstrates the intended effect of the union type is [this one](https://github.com/microsoft/TypeScript/blob/795a5c83fe55e002a98ef274e9b8c6b103ef4f74/tests/cases/conformance/salsa/moduleExportWithExportPropertyAssignment3.ts#L22-L23), where both declarations are in the JS file itself. I’m not sure if we’d expect the union type if one of those declarations were in a module augmentation. Would we expect a duplicate identifier error instead?

Reply:
> Would we expect a duplicate identifier error instead?

Probably yes, since it's a value. But to some extent we get to decide the semantics here; advanced things like module augmentations don't even necessarily need to work at all, or even have defined behaviour. (It's better if they do of course.)

Reply:
I added this test for it:

```ts
// @Filename: /test.js
module.exports = {
  a: "ok"
};

// @Filename: /index.ts
import { a } from "./test";
declare module "./test" {
  export const a: number;
}
```

And routed these `a` symbols through the `else` branch so they would go to `mergeSymbol`, thinking it would issue a duplicate identifier error. But, the JS symbol is a `Property`, which has no problem merging with the ambient `BlockScopedVariable`. As a result, the ambient `a` seems to silently overwrite the JS property `a`, at least in terms of the type baseline.

Not sure what the best approach is here. I could issue the error manually, but this is feeling very ad-hoc.